src/controllers/plugins/api/Message.cpp

```diff
@@ -232,16 +231,504 @@ std::shared_ptr<Message> messageFromTable(const sol::table &tbl)
+    {
+        return {it.current.msg, it.current.index - diff};
+    }
+
+    bool operator==(const ElementIterator &rhs) const
+    {
+        return this->current == rhs.current;
+    }
+    bool operator<(const ElementIterator &rhs) const
+    {
+        return this->current.index < rhs.current.index;
+    }
+    bool operator<=(const ElementIterator &rhs) const
+    {
+        return this->current.index <= rhs.current.index;
+    }
+    bool operator>(const ElementIterator &rhs) const
+    {
+        return this->current.index > rhs.current.index;
+    }
+    bool operator>=(const ElementIterator &rhs) const
+    {
+        return this->current.index >= rhs.current.index;
+    }
+
+    ElementRef current;
+};
+
+static_assert(std::random_access_iterator<ElementIterator>);
+
+struct MessageElements {
+    using value_type = ElementIterator::value_type;
+    using iterator = ElementIterator;
+    using size_type = size_t;
+
+    explicit MessageElements(std::shared_ptr<Message> msg)
+        : msg(std::move(msg))
+    {
+    }
+    ~MessageElements() = default;
+
+    MessageElements(const MessageElements &) = delete;
+    MessageElements(MessageElements &&) = default;
+    MessageElements &operator=(const MessageElements &) = delete;
+    MessageElements &operator=(MessageElements &&) = default;
+
+    ElementIterator begin() const
+    {
+        return {msg, 0};
+    }
+    ElementIterator end() const
+    {
+        return {msg, msg->elements.size()};
+    }
+
+    size_type size() const
+    {
+        if (!msg)
+        {
+            return 0;
+        }
+        return msg->elements.size();
+    }
+
+    // NOLINTNEXTLINE
+    size_type max_size() const
+    {
+        return this->size();  // we can't insert
+    }
+
+    bool empty() const
+    {
+        if (!msg)
+        {
+            return true;
+        }
+        return msg->elements.empty();
+    }
+
+    // NOLINTNEXTLINE
+    void push_back(ElementIterator::value_type /* v */) const
+    {
+        throw std::runtime_error("Insertion is not supported");
+    }
+
+    // NOLINTNEXTLINE
+    void erase(ElementIterator it) const
+    {
+        if (it.current.msg != this->msg || !this->msg ||
+            it.current.index >= this->msg->elements.size())
+        {
+            throw std::runtime_error("Can't erase here");
+        }
+        checkWritable(this->msg.get());
+        this->msg->elements.erase(this->msg->elements.begin() +
+                                  static_cast<ptrdiff_t>(it.current.index));
+    }
+
+    std::shared_ptr<Message> msg;
+};
+
 void createUserType(sol::table &c2)
 {
-    c2.new_usertype<Message>("Message",
-                             sol::factories([](const sol::table &tbl) {
-                                 return messageFromTable(tbl);
-                             }));
+    c2.new_usertype<ElementRef>(
+        "MessageElement", sol::no_constructor, "type",
+        sol::property([](const ElementRef &el) {
```

Comment:
do something like this to force a clang-format to insert a new line
```suggestion
        "MessageElement", sol::no_constructor, //
        "type", sol::property([](const ElementRef &el) {
```